File: Stacks/evaluate_reverse_polish_notation.py
class Solution: #O(n) time complexity, #O(n) space complexity
    def evalRPN(self, tokens: List[str]) -> int:
        res = []

        for i in range(len(tokens)):
            if tokens[i] == "+":
                res.append(res.pop(-2) + res.pop())
            elif tokens[i] == "-":
                res.append(res.pop(-2) - res.pop())
            elif tokens[i] == "*":
                res.append(res.pop(-2) * res.pop())
            elif tokens[i] == "/":
                res.append(int(res.pop(-2) / res.pop())) #int() function needed to truncate dividends towards zero
            else:
                res.append(int(tokens[i]))

        return res[0]

File: Stacks/evaluate_reverse_polish_notation.js
/*
https://leetcode.com/problems/evaluate-reverse-polish-notation/description/
*/

class Solution {
    /**
     * @param {string[]} tokens
     * @return {number}
     * Time complexity O(n), Space complexoty O(n)
     */
    evalRPN(tokens) {
        const stack = []
        for(var token of tokens){
            if(token === "+"){
                let num1 = stack.pop()
                let num2 = stack.pop()
                stack.push(num2 + num1)
            }else if(token === "-"){
                let num1 = stack.pop()
                let num2 = stack.pop()
                stack.push(num2 - num1)
            }else if(token === "/"){
                let num1 = stack.pop()
                let num2 = stack.pop()
                stack.push(Math.trunc(num2 / num1))
            }else if(token === "*"){
                let num1 = stack.pop()
                let num2 = stack.pop()
                stack.push(num2 * num1)
            }else{
                stack.push(Number(token))
            }
        }
        return stack.pop()
    }
}
